lib/bookingsync_portal/booking_map.rb

```diff
@@ -0,0 +1,99 @@
+class BookingsyncPortal::BookingMap
+  class InvalidLength < StandardError
+  end
+
+  class << self
+    # Diff maps
+    #
+    # @param source [String] The source map (expecting a map of `1` and `0` only)
+    # @param destination [String] The destination map (expecting a map of `1` and `0` only)
+    # @return [String] A map with the changes between both maps:
+    #   _ - Identical
+    #   1 - Create (present on source only)
+    #   0 - Destroy (present on destination only)
+    def diff(source, destination)
+      raise InvalidLength if source.size != destination.size
+
+      new_result = "_" * source.size
+      source.chars.each_with_index do |char, index|
+        new_result[index] = char if destination[index] != char
+      end
+
+      new_result
+    end
+
+    # Extract ranges
+    #
+    # Returns all the ranges with their statuses.
+    # The `_` is considered as empty and won't be returned as a range.
+    # @param map [String] A map String
+    # @param from [Date] The day that the map starts
+    # @return [Array<Hash>] Returns an Array of Hashes with all the ranges
+    def extract_ranges(map, from)
+      ranges = []
+      previous_char = nil
+      range = nil
+      map_length = map.length
+      map.chars.each_with_index do |char, index|
+        if char != previous_char && char != "_" # finished
+          ranges << range if range
+
+          range = {
+            start_at: from.advance(days: index),
+            end_at: from.advance(days: index + 1)
+          }.merge(status: char)
+        elsif char != "_" # continuing
+          range[:end_at] = from.advance(days: index + 1)
+        end
+
+        if index == map_length - 1
+          ranges << range if range
+        end
+
+        previous_char = char
+      end
+
+      ranges
+    end
+  end
+
+  attr_reader :bookings, :statuses, :from, :length
+
+  def initialize(bookings, options = {})
+    @from = options.fetch(:from) { Date.today.beginning_of_month }
+    @length = options.fetch(:length) { 1096 }
+    @statuses = options.fetch(:statuses) { {
+      available:    '0',
+      tentative:    '0',
+      booked:       '1',
+      unavailable:  '1'
+    } }
+    @bookings = bookings_for_map(bookings: bookings, from: @from, length: @length)
+  end
+
+  def map
+    map = statuses[:available].to_s * length
+
+    bookings.each do |booking|
+      (booking.start_at.to_date...booking.end_at.to_date).each do |date|
+        index = (date - from.to_date).to_i
+
+        if index >= 0 && index < length
+          map[index] = statuses[booking.status.downcase.to_sym]
+        end
+      end
+    end
+    map
```

Comment:
I think this algorithm is a bit too cryptic with indexes, like low level C-style programming. How about using hash with dates as keys and statuses as values? Would be much more readable then and easier to debug.

It would require to initialize hash with range from bookings lowest start_at date and highest end_at date and some default value (probably as available) date and iterate over the dates ranges just like now, but instead of dealing with indexes it would be dealing with dates in hash. How does it sound?